source/endpoints_ruleset.c

```diff
@@ -0,0 +1,939 @@
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract parameter documentation.");
+            goto error_clean_up;
+        }
+        rule.documentation = aws_string_new_from_cursor(wrapper->allocator, &documentation_cur);
+    }
+
+    aws_array_list_push_back(wrapper->array, &rule);
+
+    return AWS_OP_SUCCESS;
+
+error_clean_up:
+    aws_endpoints_rule_clean_up(&rule);
+    return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+}
+
+static int s_init_ruleset_from_json(
+    struct aws_allocator *allocator,
+    struct aws_endpoints_ruleset *ruleset,
+    struct aws_byte_cursor json) {
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(ruleset);
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&json));
+
+    /*
+     * TODO: this could be more efficient. currently user allocates mem for json
+     * string, we create a copy in json parser and then string fields create
+     * another copy.
+     */
+
+    struct aws_json_value *root = aws_json_value_new_from_string(allocator, json);
+
+    if (root == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to parse provided string as json.");
+        return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+    }
+
+    struct aws_byte_cursor version_cur;
+    struct aws_json_value *version_node = aws_json_value_get_from_object(root, aws_byte_cursor_from_c_str("version"));
+    if (version_node == NULL || aws_json_value_get_string(version_node, &version_cur)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract version.");
+        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_UNSUPPORTED_RULESET);
+        goto error_clean_up;
+    }
+
+#ifdef VERSION_CHECK /* TODO: samples are currently inconsistent with versions. skip check for now */
+    if (aws_byte_cursor_eq_c_str(&version_cur, &s_supported_version)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Unsupported ruleset version.");
+        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_UNSUPPORTED_RULESET);
+        goto error_clean_up;
+    }
+#endif
+
+    ruleset->version = aws_string_new_from_cursor(allocator, &version_cur);
+
+    struct aws_byte_cursor service_id_cur;
+    struct aws_json_value *service_id_node =
+        aws_json_value_get_from_object(root, aws_byte_cursor_from_c_str("serviceId"));
+
+    if (service_id_node != NULL) {
+        if (aws_json_value_get_string(service_id_node, &service_id_cur)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract serviceId.");
+            aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_UNSUPPORTED_RULESET);
+            goto error_clean_up;
+        }
+        ruleset->service_id = aws_string_new_from_cursor(allocator, &service_id_cur);
+    }
+
+    ruleset->parameters = aws_mem_acquire(allocator, sizeof(struct aws_hash_table));
+    aws_hash_table_init(
+        ruleset->parameters,
```

Comment:
trivial:  might be simpler to store the hashtable by-value instead of on the heap? seems like it always exists, so why not?